File: packages/go/cairn-p2p/crypto/noise_test.go
package crypto

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func runHandshake(t *testing.T, pakeSecret []byte) (*HandshakeResult, *HandshakeResult) {
	t.Helper()
	aliceID, err := GenerateIdentity()
	require.NoError(t, err)
	bobID, err := GenerateIdentity()
	require.NoError(t, err)
	return runHandshakeWithIdentities(t, aliceID, bobID, pakeSecret)
}

func runHandshakeWithIdentities(t *testing.T, aliceID, bobID *IdentityKeypair, pakeSecret []byte) (*HandshakeResult, *HandshakeResult) {
	t.Helper()

	initiator := NewNoiseXX(RoleInitiator, aliceID, pakeSecret)
	responder := NewNoiseXX(RoleResponder, bobID, pakeSecret)

	// Initiator sends msg1
	out1, err := initiator.Step(nil)
	require.NoError(t, err)
	require.NotNil(t, out1.Message)
	require.Nil(t, out1.Complete)

	// Responder receives msg1, sends msg2
	out2, err := responder.Step(out1.Message)
	require.NoError(t, err)
	require.NotNil(t, out2.Message)
	require.Nil(t, out2.Complete)

	// Initiator receives msg2, sends msg3
	out3, err := initiator.Step(out2.Message)
	require.NoError(t, err)
	require.NotNil(t, out3.Message)

	// Get initiator result
	initResult, err := initiator.Result()
	require.NoError(t, err)

	// Responder receives msg3
	out4, err := responder.Step(out3.Message)
	require.NoError(t, err)
	require.Nil(t, out4.Message)
	require.NotNil(t, out4.Complete)

	return initResult, out4.Complete
}

func TestFullHandshakeProducesMatchingSessionKeys(t *testing.T) {
	initResult, respResult := runHandshake(t, nil)
	assert.Equal(t, initResult.SessionKey, respResult.SessionKey)
}

func TestHandshakeRevealsRemoteStaticKeys(t *testing.T) {
	aliceID, err := GenerateIdentity()
	require.NoError(t, err)
	bobID, err := GenerateIdentity()
	require.NoError(t, err)

	alicePub := aliceID.PublicKey()
	bobPub := bobID.PublicKey()

	initResult, respResult := runHandshakeWithIdentities(t, aliceID, bobID, nil)

	assert.Equal(t, bobPub, initResult.RemoteStatic)
	assert.Equal(t, alicePub, respResult.RemoteStatic)
}

func TestHandshakeTranscriptHashesMatch(t *testing.T) {
	initResult, respResult := runHandshake(t, nil)
	assert.Equal(t, initResult.TranscriptHash, respResult.TranscriptHash)
}

func TestDifferentHandshakesProduceDifferentSessionKeys(t *testing.T) {
	result1, _ := runHandshake(t, nil)
	result2, _ := runHandshake(t, nil)
	assert.NotEqual(t, result1.SessionKey, result2.SessionKey)
}

func TestHandshakeWithPakeSecret(t *testing.T) {
	pake := make([]byte, 32)
	for i := range pake {
		pake[i] = 42
	}
	initResult, respResult := runHandshake(t, pake)
	assert.Equal(t, initResult.SessionKey, respResult.SessionKey)
}

func TestMismatchedPakeSecretsFail(t *testing.T) {
	aliceID, err := GenerateIdentity()
	require.NoError(t, err)
	bobID, err := GenerateIdentity()
	require.NoError(t, err)

	pake1 := make([]byte, 32)
	pake1[0] = 1
	pake2 := make([]byte, 32)
	pake2[0] = 2

	initiator := NewNoiseXX(RoleInitiator, aliceID, pake1)
	responder := NewNoiseXX(RoleResponder, bobID, pake2)

	out1, err := initiator.Step(nil)
	require.NoError(t, err)

	out2, err := responder.Step(out1.Message)
	require.NoError(t, err)

	out3, err := initiator.Step(out2.Message)
	require.NoError(t, err)

	// Responder should fail to decrypt msg3
	_, err = responder.Step(out3.Message)
	assert.Error(t, err)
}

func TestMsg1WrongLengthRejected(t *testing.T) {
	bobID, err := GenerateIdentity()
	require.NoError(t, err)
	responder := NewNoiseXX(RoleResponder, bobID, nil)

	_, err = responder.Step([]byte{0, 1, 2, 3})
	assert.Error(t, err)
}

func TestMsg2TooShortRejected(t *testing.T) {
	aliceID, err := GenerateIdentity()
	require.NoError(t, err)
	initiator := NewNoiseXX(RoleInitiator, aliceID, nil)

	out1, err := initiator.Step(nil)
	require.NoError(t, err)

	// Pass a truncated message 2
	_, err = initiator.Step(out1.Message[:10])
	assert.Error(t, err)
}

func TestTamperedMsg2Rejected(t *testing.T) {
	aliceID, err := GenerateIdentity()
	require.NoError(t, err)
	bobID, err := GenerateIdentity()
	require.NoError(t, err)

	initiator := NewNoiseXX(RoleInitiator, aliceID, nil)
	responder := NewNoiseXX(RoleResponder, bobID, nil)

	out1, err := initiator.Step(nil)
	require.NoError(t, err)

	out2, err := responder.Step(out1.Message)
	require.NoError(t, err)

	// Tamper with encrypted portion
	if len(out2.Message) > 40 {
		out2.Message[40] ^= 0xFF
	}

	_, err = initiator.Step(out2.Message)
	assert.Error(t, err)
}

func TestTamperedMsg3Rejected(t *testing.T) {
	aliceID, err := GenerateIdentity()
	require.NoError(t, err)
	bobID, err := GenerateIdentity()
	require.NoError(t, err)

	initiator := NewNoiseXX(RoleInitiator, aliceID, nil)
	responder := NewNoiseXX(RoleResponder, bobID, nil)

	out1, err := initiator.Step(nil)
	require.NoError(t, err)

	out2, err := responder.Step(out1.Message)
	require.NoError(t, err)

	out3, err := initiator.Step(out2.Message)
	require.NoError(t, err)

	out3.Message[0] ^= 0xFF

	_, err = responder.Step(out3.Message)
	assert.Error(t, err)
}

func TestOutOfOrderStepRejected(t *testing.T) {
	aliceID, err := GenerateIdentity()
	require.NoError(t, err)
	initiator := NewNoiseXX(RoleInitiator, aliceID, nil)

	_, err = initiator.Step([]byte{0})
	assert.Error(t, err)
}

func TestResponderRejectsNoInput(t *testing.T) {
	bobID, err := GenerateIdentity()
	require.NoError(t, err)
	responder := NewNoiseXX(RoleResponder, bobID, nil)

	_, err = responder.Step(nil)
	assert.Error(t, err)
}

func TestStepAfterCompleteRejected(t *testing.T) {
	aliceID, err := GenerateIdentity()
	require.NoError(t, err)
	bobID, err := GenerateIdentity()
	require.NoError(t, err)

	initiator := NewNoiseXX(RoleInitiator, aliceID, nil)
	responder := NewNoiseXX(RoleResponder, bobID, nil)

	out1, err := initiator.Step(nil)
	require.NoError(t, err)
	out2, err := responder.Step(out1.Message)
	require.NoError(t, err)
	out3, err := initiator.Step(out2.Message)
	require.NoError(t, err)

	out4, err := responder.Step(out3.Message)
	require.NoError(t, err)
	require.NotNil(t, out4.Complete)

	_, err = responder.Step(nil)
	assert.Error(t, err)
}

func TestMsg1Is32Bytes(t *testing.T) {
	aliceID, err := GenerateIdentity()
	require.NoError(t, err)
	initiator := NewNoiseXX(RoleInitiator, aliceID, nil)

	out, err := initiator.Step(nil)
	require.NoError(t, err)
	assert.Len(t, out.Message, 32)
}

// --- SAS tests ---

func TestSASDerivableFromHandshake(t *testing.T) {
	initResult, respResult := runHandshake(t, nil)

	initSAS, err := NumericSAS(initResult.TranscriptHash)
	require.NoError(t, err)
	respSAS, err := NumericSAS(respResult.TranscriptHash)
	require.NoError(t, err)

	assert.Equal(t, initSAS, respSAS)
}

func TestEmojiSASMatchesBetweenPeers(t *testing.T) {
	initResult, respResult := runHandshake(t, nil)

	initEmoji, err := EmojiSAS(initResult.TranscriptHash)
	require.NoError(t, err)
	respEmoji, err := EmojiSAS(respResult.TranscriptHash)
	require.NoError(t, err)

	assert.Equal(t, initEmoji, respEmoji)
}

func TestNumericSASFormat(t *testing.T) {
	hash := [32]byte{42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
		42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42, 42}
	sas, err := NumericSAS(hash)
	require.NoError(t, err)
	assert.Len(t, sas, 6)
	for _, c := range sas {
		assert.True(t, c >= '0' && c <= '9')
	}
}

func TestNumericSASIsDeterministic(t *testing.T) {
	hash := [32]byte{99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99,
		99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99}
	sas1, err := NumericSAS(hash)
	require.NoError(t, err)
	sas2, err := NumericSAS(hash)
	require.NoError(t, err)
	assert.Equal(t, sas1, sas2)
}

func TestDifferentTranscriptsProduceDifferentSAS(t *testing.T) {
	hash1 := [32]byte{1}
	hash2 := [32]byte{2}
	sas1, err := NumericSAS(hash1)
	require.NoError(t, err)
	sas2, err := NumericSAS(hash2)
	require.NoError(t, err)
	assert.NotEqual(t, sas1, sas2)
}

func TestEmojiSASReturns4Entries(t *testing.T) {
	hash := [32]byte{42}
	emojis, err := EmojiSAS(hash)
	require.NoError(t, err)
	assert.Len(t, emojis, 4)
}

func TestEmojiSASIsDeterministic(t *testing.T) {
	hash := [32]byte{99}
	e1, err := EmojiSAS(hash)
	require.NoError(t, err)
	e2, err := EmojiSAS(hash)
	require.NoError(t, err)
	assert.Equal(t, e1, e2)
}

func TestEmojiSASEntriesAreFromTable(t *testing.T) {
	hash := [32]byte{77}
	emojis, err := EmojiSAS(hash)
	require.NoError(t, err)
	for _, emoji := range emojis {
		found := false
		for _, e := range EmojiTable {
			if emoji == e {
				found = true
				break
			}
		}
		assert.True(t, found, "emoji %q not found in table", emoji)
	}
}

File: packages/ts/cairn-p2p/tests/unit/noise.test.ts
import { describe, it, expect } from 'vitest';
import { NoiseXXHandshake } from '../../src/crypto/noise.js';
import type { HandshakeResult, StepOutput } from '../../src/crypto/noise.js';
import { IdentityKeypair } from '../../src/crypto/identity.js';
import { deriveNumericSas, deriveEmojiSas, EMOJI_TABLE } from '../../src/crypto/sas.js';
import { Spake2 } from '../../src/crypto/spake2.js';
import { CairnError } from '../../src/errors.js';

/** Run a complete Noise XX handshake and return both results. */
async function runHandshake(
  pakeSecret?: Uint8Array,
): Promise<[HandshakeResult, HandshakeResult]> {
  const alice = await IdentityKeypair.generate();
  const bob = await IdentityKeypair.generate();
  return runHandshakeWithIdentities(alice, bob, pakeSecret);
}

async function runHandshakeWithIdentities(
  alice: IdentityKeypair,
  bob: IdentityKeypair,
  pakeSecret?: Uint8Array,
): Promise<[HandshakeResult, HandshakeResult]> {
  const initiator = new NoiseXXHandshake('initiator', alice, pakeSecret);
  const responder = new NoiseXXHandshake('responder', bob, pakeSecret);

  // Initiator sends msg1
  const out1 = initiator.step();
  expect(out1.type).toBe('send_message');
  const msg1 = (out1 as { type: 'send_message'; data: Uint8Array }).data;

  // Responder receives msg1, sends msg2
  const out2 = responder.step(msg1);
  expect(out2.type).toBe('send_message');
  const msg2 = (out2 as { type: 'send_message'; data: Uint8Array }).data;

  // Initiator receives msg2, sends msg3
  const out3 = initiator.step(msg2);
  expect(out3.type).toBe('send_message');
  const msg3 = (out3 as { type: 'send_message'; data: Uint8Array }).data;

  // Initiator result is available via getResult()
  const initiatorResult = initiator.getResult();

  // Responder receives msg3 -> complete
  const out4 = responder.step(msg3);
  expect(out4.type).toBe('complete');
  const responderResult = (out4 as { type: 'complete'; result: HandshakeResult }).result;

  return [initiatorResult, responderResult];
}

describe('Noise XX Handshake', () => {
  it('produces matching session keys', async () => {
    const [initResult, respResult] = await runHandshake();
    expect(initResult.sessionKey).toEqual(respResult.sessionKey);
  });

  it('reveals remote static keys', async () => {
    const alice = await IdentityKeypair.generate();
    const bob = await IdentityKeypair.generate();
    const alicePub = alice.publicKey();
    const bobPub = bob.publicKey();

    const [initResult, respResult] = await runHandshakeWithIdentities(alice, bob);

    expect(initResult.remoteStatic).toEqual(bobPub);
    expect(respResult.remoteStatic).toEqual(alicePub);
  });

  it('transcript hashes match', async () => {
    const [initResult, respResult] = await runHandshake();
    expect(initResult.transcriptHash).toEqual(respResult.transcriptHash);
  });

  it('different handshakes produce different session keys', async () => {
    const [result1] = await runHandshake();
    const [result2] = await runHandshake();
    expect(result1.sessionKey).not.toEqual(result2.sessionKey);
  });

  it('message 1 is 32 bytes (ephemeral public key)', async () => {
    const alice = await IdentityKeypair.generate();
    const initiator = new NoiseXXHandshake('initiator', alice);
    const out = initiator.step();
    expect(out.type).toBe('send_message');
    expect((out as { type: 'send_message'; data: Uint8Array }).data.length).toBe(32);
  });

  it('session key is 32 bytes', async () => {
    const [initResult] = await runHandshake();
    expect(initResult.sessionKey.length).toBe(32);
  });

  it('transcript hash is 32 bytes', async () => {
    const [initResult] = await runHandshake();
    expect(initResult.transcriptHash.length).toBe(32);
  });
});

describe('Noise XX with PAKE secret', () => {
  it('completes with matching PAKE secrets', async () => {
    const pake = new Uint8Array(32).fill(42);
    const [initResult, respResult] = await runHandshake(pake);
    expect(initResult.sessionKey).toEqual(respResult.sessionKey);
  });

  it('mismatched PAKE secrets cause failure', async () => {
    const alice = await IdentityKeypair.generate();
    const bob = await IdentityKeypair.generate();

    const initiator = new NoiseXXHandshake('initiator', alice, new Uint8Array(32).fill(1));
    const responder = new NoiseXXHandshake('responder', bob, new Uint8Array(32).fill(2));

    const msg1 = (initiator.step() as { type: 'send_message'; data: Uint8Array }).data;
    const msg2 = (responder.step(msg1) as { type: 'send_message'; data: Uint8Array }).data;
    const msg3 = (initiator.step(msg2) as { type: 'send_message'; data: Uint8Array }).data;

    // Responder should fail to decrypt msg3 because PAKE secrets differ
    expect(() => responder.step(msg3)).toThrow(CairnError);
  });
});

describe('Noise XX error handling', () => {
  it('initiator rejects input at start', async () => {
    const alice = await IdentityKeypair.generate();
    const initiator = new NoiseXXHandshake('initiator', alice);
    expect(() => initiator.step(new Uint8Array(32))).toThrow(CairnError);
  });

  it('responder rejects no input', async () => {
    const bob = await IdentityKeypair.generate();
    const responder = new NoiseXXHandshake('responder', bob);
    expect(() => responder.step()).toThrow(CairnError);
  });

  it('message 1 wrong length rejected', async () => {
    const bob = await IdentityKeypair.generate();
    const responder = new NoiseXXHandshake('responder', bob);
    expect(() => responder.step(new Uint8Array(16))).toThrow(CairnError);
  });

  it('message 2 too short rejected', async () => {
    const alice = await IdentityKeypair.generate();
    const bob = await IdentityKeypair.generate();

    const initiator = new NoiseXXHandshake('initiator', alice);
    const responder = new NoiseXXHandshake('responder', bob);

    const msg1 = (initiator.step() as { type: 'send_message'; data: Uint8Array }).data;
    responder.step(msg1);

    expect(() => initiator.step(new Uint8Array(10))).toThrow(CairnError);
  });

  it('message 3 too short rejected', async () => {
    const alice = await IdentityKeypair.generate();
    const bob = await IdentityKeypair.generate();

    const initiator = new NoiseXXHandshake('initiator', alice);
    const responder = new NoiseXXHandshake('responder', bob);

    const msg1 = (initiator.step() as { type: 'send_message'; data: Uint8Array }).data;
    responder.step(msg1);

    expect(() => responder.step(new Uint8Array(5))).toThrow(CairnError);
  });

  it('tampered message 2 rejected', async () => {
    const alice = await IdentityKeypair.generate();
    const bob = await IdentityKeypair.generate();

    const initiator = new NoiseXXHandshake('initiator', alice);
    const responder = new NoiseXXHandshake('responder', bob);

    const msg1 = (initiator.step() as { type: 'send_message'; data: Uint8Array }).data;
    const msg2 = (responder.step(msg1) as { type: 'send_message'; data: Uint8Array }).data;

    // Tamper with encrypted portion
    if (msg2.length > 40) msg2[40] ^= 0xff;
    expect(() => initiator.step(msg2)).toThrow(CairnError);
  });

  it('tampered message 3 rejected', async () => {
    const alice = await IdentityKeypair.generate();
    const bob = await IdentityKeypair.generate();

    const initiator = new NoiseXXHandshake('initiator', alice);
    const responder = new NoiseXXHandshake('responder', bob);

    const msg1 = (initiator.step() as { type: 'send_message'; data: Uint8Array }).data;
    const msg2 = (responder.step(msg1) as { type: 'send_message'; data: Uint8Array }).data;
    const msg3 = (initiator.step(msg2) as { type: 'send_message'; data: Uint8Array }).data;

    msg3[0] ^= 0xff;
    expect(() => responder.step(msg3)).toThrow(CairnError);
  });

  it('step after complete rejected', async () => {
    const [, respResult] = await runHandshake();
    // Create a new completed handshake and test further steps
    const alice = await IdentityKeypair.generate();
    const bob = await IdentityKeypair.generate();

    const initiator = new NoiseXXHandshake('initiator', alice);
    const responder = new NoiseXXHandshake('responder', bob);

    const msg1 = (initiator.step() as { type: 'send_message'; data: Uint8Array }).data;
    const msg2 = (responder.step(msg1) as { type: 'send_message'; data: Uint8Array }).data;
    const msg3 = (initiator.step(msg2) as { type: 'send_message'; data: Uint8Array }).data;
    responder.step(msg3);

    expect(() => responder.step()).toThrow(CairnError);
  });
});

describe('SAS derivation', () => {
  it('numeric SAS matches between peers', async () => {
    const [initResult, respResult] = await runHandshake();
    const initSas = deriveNumericSas(initResult.transcriptHash);
    const respSas = deriveNumericSas(respResult.transcriptHash);
    expect(initSas).toBe(respSas);
  });

  it('emoji SAS matches between peers', async () => {
    const [initResult, respResult] = await runHandshake();
    const initEmoji = deriveEmojiSas(initResult.transcriptHash);
    const respEmoji = deriveEmojiSas(respResult.transcriptHash);
    expect(initEmoji).toEqual(respEmoji);
  });

  it('numeric SAS is 6 digits', () => {
    const hash = new Uint8Array(32).fill(42);
    const sas = deriveNumericSas(hash);
    expect(sas.length).toBe(6);
    expect(sas).toMatch(/^\d{6}$/);
  });

  it('numeric SAS is zero-padded', () => {
    // We can't easily force a specific result, but let's verify format
    for (let i = 0; i < 10; i++) {
      const hash = new Uint8Array(32).fill(i);
      const sas = deriveNumericSas(hash);
      expect(sas.length).toBe(6);
      expect(sas).toMatch(/^\d{6}$/);
    }
  });

  it('numeric SAS is deterministic', () => {
    const hash = new Uint8Array(32).fill(99);
    expect(deriveNumericSas(hash)).toBe(deriveNumericSas(hash));
  });

  it('different transcripts produce different SAS', () => {
    const hash1 = new Uint8Array(32).fill(1);
    const hash2 = new Uint8Array(32).fill(2);
    expect(deriveNumericSas(hash1)).not.toBe(deriveNumericSas(hash2));
  });

  it('emoji SAS returns 4 entries', () => {
    const hash = new Uint8Array(32).fill(42);
    const emojis = deriveEmojiSas(hash);
    expect(emojis.length).toBe(4);
  });

  it('emoji SAS is deterministic', () => {
    const hash = new Uint8Array(32).fill(99);
    expect(deriveEmojiSas(hash)).toEqual(deriveEmojiSas(hash));
  });

  it('emoji SAS entries are from table', () => {
    const hash = new Uint8Array(32).fill(77);
    const emojis = deriveEmojiSas(hash);
    for (const emoji of emojis) {
      expect(EMOJI_TABLE).toContain(emoji);
    }
  });

  it('emoji table has 64 entries', () => {
    expect(EMOJI_TABLE.length).toBe(64);
  });
});

describe('SPAKE2', () => {
  it('same password produces matching keys', () => {
    const password = new TextEncoder().encode('test-password-42');
    const alice = Spake2.startA(password);
    const bob = Spake2.startB(password);

    const aliceKey = alice.finish(bob.outboundMsg);
    const bobKey = bob.finish(alice.outboundMsg);

    expect(aliceKey).toEqual(bobKey);
  });

  it('different passwords produce different keys', () => {
    const alice = Spake2.startA(new TextEncoder().encode('password-1'));
    const bob = Spake2.startB(new TextEncoder().encode('password-2'));

    const aliceKey = alice.finish(bob.outboundMsg);
    const bobKey = bob.finish(alice.outboundMsg);

    expect(aliceKey).not.toEqual(bobKey);
  });

  it('output key is 32 bytes', () => {
    const password = new TextEncoder().encode('test');
    const alice = Spake2.startA(password);
    const bob = Spake2.startB(password);
    const key = alice.finish(bob.outboundMsg);
    expect(key.length).toBe(32);
  });

  it('outbound message is 33 bytes (side prefix + compressed Ed25519 point)', () => {
    const password = new TextEncoder().encode('test');
    const alice = Spake2.startA(password);
    expect(alice.outboundMsg.length).toBe(33);
  });

  it('different sessions produce different outbound messages', () => {
    const password = new TextEncoder().encode('same-password');
    const a1 = Spake2.startA(password);
    const a2 = Spake2.startA(password);
    expect(a1.outboundMsg).not.toEqual(a2.outboundMsg);
  });

  it('SPAKE2 key integrates with Noise XX as PAKE secret', async () => {
    const password = new TextEncoder().encode('pairing-pin-42');

    // Both sides run SPAKE2 to derive a shared key
    const spakeA = Spake2.startA(password);
    const spakeB = Spake2.startB(password);

    const pakeKeyA = spakeA.finish(spakeB.outboundMsg);
    const pakeKeyB = spakeB.finish(spakeA.outboundMsg);

    expect(pakeKeyA).toEqual(pakeKeyB);

    // Use the SPAKE2 key as PAKE secret in Noise XX
    const alice = await IdentityKeypair.generate();
    const bob = await IdentityKeypair.generate();

    const [initResult, respResult] = await runHandshakeWithIdentities(alice, bob, pakeKeyA);

    expect(initResult.sessionKey).toEqual(respResult.sessionKey);
  });
});
